### ISSUE TITLE
[Bug] Resurrect `pandera` support for the good of `pandas` :panda_face: 

### ISSUE THREAD

**Issue by komodovaran** [beartype/beartype#337]
The following does not work, despite `pandas-stubs` claiming some sort of PEP-compliance:

```python
from __future__ import annotations

from typing import TYPE_CHECKING

import pandas as pd
from beartype import beartype


@beartype
def some_func() -> pd.Series[int]:
    return pd.Series([1, 2, 3])

some_func()

if TYPE_CHECKING:
    reveal_type(some_func())  # Mypy: Revealed type is "pandas.core.series.Series[builtins.int]"
```

This raises
```python
beartype.roar.BeartypeDecorHintForwardRefException: Function __main__.some_func() return stringified type hint 'pd.Series[int]' syntactically invalid (i.e., TypeError("type 'Series' is not subscriptable")).
```

Maybe this is a problem of the `pd.Series` type not having a `__getitem__`? If we then add that, based on this completely random comment I found on the internet:
https://stackoverflow.com/questions/77697685/python-generic-type-annotation-for-pandas-series#comment136990299_77697685

like so:
```python
pd.Series.__class_getitem__ = lambda cls, *_: cls
```

But now I've unlocked a whole new level of pain:

```python
beartype.roar.BeartypeCallHintReturnViolation: Function __main__.some_func() return "0    1
1    2
2    3
dtype: int64" violates type hint <class 'int'>, as <protocol "pandas.core.series.Series"> "0    1
1    2
2    3
dtype: int64" not instance of int.
```

And changing the generic form `int` to, say, `np.int` or whatever will now upset `mypy` as well (which expects `int`). `beartype` will not be satisfied by any attempts at salvaging the hints.

**Comment by leycec** [comment]
*...heh.* `pandas-stubs`, huh? **Yeah. About that.**

The First Rule of Beartype Club is that nobody knows how to implement type hints. The Second Rule of Beartype Club is that two projects actually *do* know how to implement type hints. They are unsung heroes whose names shall be entered into the Akashic Records housed at Valhalla, Asgard:

* @patrick-kidger's [`jaxtyping`](https://docs.kidger.site/jaxtyping), covering JAX, NumPy, PyTorch, and TensorFlow tensors (i.e., multidimensional arrays).
* @cosmicBboy's [`pandera`](https://pandera.readthedocs.io), covering pandas `DataFrame`s.

That's it. All other type hints should be assumed to be insane, because they are. This includes type hints published by:

* The standard `typing` module, which @beartype only supports because it has to. The standard `typing` module is still insane, though. CPython devs have *no* idea what they're doing at all. Truth or dare? Mostly truth.
* The `jax.typing` subpackage. *Ditto.*
* The `numpy.typing` subpackage. *Ditto.*
* The `pandas-stubs` package. *Ditto.*
* And everybody else, too.

Thankfully, you're in luck! @beartype officially supports `pandera`. I may be crabby, but there's no reason you need to suffer:

```python
from __future__ import annotations

from pandas import Series
from pandera.dtypes import Int64
from pandera.typing import Series as SeriesBusiness  # <-- get it, "series business"? ...i'll stop now
from beartype import beartype

@beartype
def some_func(dataframe: SeriesBusiness[Int64]) -> None:
    print(dataframe)

some_func(Series([1, 2, 3]))
some_func("Pretty sure this ain't a dataframe.")
```

...which prints:

```python
  from pandas import Series
0    1
1    2
2    3
dtype: int64
Traceback (most recent call last):
  File "/home/leycec/tmp/mopy.py", line 15, in <module>
    some_func("Pretty sure this ain't a dataframe.")
  File "<@beartype(__main__.some_func) at 0x7f8766fbc4a0>", line 30, in some_func
beartype.roar.BeartypeCallHintParamViolation: Function __main__.some_func() parameter
dataframe="Pretty sure this ain't a dataframe." violates type hint
pandera.typing.pandas.Series[pandera.dtypes.Int64], as str "Pretty sure this ain't a
dataframe." not instance of <protocol "pandas.core.series.Series">.
```

See also [@beartype's official FAQ entry on pandas type-checking](https://beartype.readthedocs.io/en/latest/faq/#pandas-data-frames). Now, if everyone will excuse me, I must some crab apples sprinkled with 20 year-old cinnamon to feel better. I didn't know that 20 year-old cinnamon could still be edible. But... *it can.*

<sup>**True story:** [in Denmark, "friends" and family publicly parade you through the streets covered in cinnamon if you're still unmarried on your 25th birthday. As if that wasn't bad enough, the punishment even intensifies as you age.](https://www.businessinsider.com/why-singles-in-denmark-get-covered-in-cinnamon-2018-2) Thanks, Business Insider. You have taught me I narrowly dodged a cultural bullet.</sup>

**Comment by leycec** [comment]
**Oh, Gods.** I double-checked this and... *something is busted.* Looks like `pandera` "improved" itself at some point and I failed to notice. This is bad. Although @beartype *does* still nominally support `pandera` type hints, @beartype no longer appears to correctly type-check pandas `DataFrame`s against `pandera` type hints. `pandera` unit tests in the @beartype test suite are currently passing – but they shouldn't be. Badness continues.

Let's revive this issue under a new header: "Resurrect `pandera` support for the good of pandas"! :panda_face:

**Comment by cosmicBboy** [comment]
uh oh, curious to learn what's causing this issue pandera-side

**Comment by leycec** [comment]
It's probably us – *but it might be you.* But it's probably us. Regardless of who it is, thanks so much for chiming in so readily. You're the pandas chef's kiss, @cosmicBboy. :panda_face:

Since you're up and about, would you mind you confirm that the following minimal-reproducible example is expected to pass rather than raise an exception? Something doesn't seem quite right. We've cited something similar in [@beartype's official FAQ entry promoting `pandera`](https://beartype.readthedocs.io/en/latest/faq/#pandas-data-frames). So, I'd *really* like to get this right:

```python
from pandas import Series
from pandera import check_types as pandera_check_types
from pandera.dtypes import Int64
from pandera.typing import Series as SeriesBusiness  # <-- get it, "series business"? ...i'll stop now

@pandera_check_types
def some_func(dataframe: SeriesBusiness[Int64]) -> None:
    print(dataframe)

some_func(Series([1, 2, 3]))  # <-- this passes. *GOOD*
some_func(Series(['ok', 'ko', 'yo']))  # <-- this also passes. that can't be right? *BAD*
some_func(Series([1.2, 2.3, 3.4]))  # <----- this also passes. that can't be right? *BAD*
```

...which prints:

```python
0    1
1    2
2    3
dtype: int64
0    ok
1    ko
2    yo
dtype: object
0    1.2
1    2.3
2    3.4
dtype: float64
```

Basically, everything passes – but I kinda only expected the first call to `some_func()` to pass. Clearly, I'm confused.

There's just a veritable *ton* of different `pandera` decorators. That's not a bad thing *per say.* But I'm unclear which one I should be calling. `@pandera.check_types` seems right, but maybe I want `@pandera.check_input` or `@pandera.check` instead or *whatevahs*? Also, how actually *do* you type-check pandas `Series` objects? I couldn't quite tell from the documentation. I assumed `pandera.typing.Series[int]` or *whatevahs* would do the trick, but now I'm getting no treats.

Clearly, I have no idea what I'm doing here. Help a @beartype bro help `pandera`.

**Comment by leycec** [comment]
@komodovaran: In theory, [pandera](https://github.com/unionai-oss/pandera) is the answer. In practice, let's see how unionai-oss/pandera#1527 pans out. :smiling_face_with_tear: 

@cosmicBboy: Because I'm clueless, I've upstreamed this into your court at unionai-oss/pandera#1527. I'm out of my league and playing with a half-empty deck. *It's not a great feeling.*
